### PR TITLE
Modstore: show an ui error if retrieving failed

### LINKED ISSUE #3042: Cannot load mod list
`2015-08-13 01:51:29: ERROR[AsyncWorkerThread_1]: https://forum.minetest.net/mmdb/mods/ not found (HTTP response code said error) (response code 404)`

Commit: 738fbc6

### PR DESCRIPTION
**Pull request by est31**
Fixes #3042.

### CODE REVIEW

**Comment by VanessaE** [comment]
Seems to work; In the UI I get "Mod list download failed: HTTP request failed with code 404" when I attempt to connect to the online mod repo.  Can't say it's all that friendly to users, but it's enough.  It might be better to what some websites do, i.e.:

Oops, the mod list failed to download.
Error code 404, sorry about that!

File: src/script/lua_api/l_mainmenu.cpp [hunk 1]
```diff
@@ -331,7 +331,12 @@ int ModApiMainMenu::l_get_modstore_details(lua_State *L)
 		url.erase(idpos,1);
 		url.insert(idpos,modid);
 
-		details = getModstoreUrl(url);
+		try {
+			details = getModstoreUrl(url);
+		} catch (JsonFetchException &e) {
+			lua_pushnil(L);
+			return 1;
```
Comment (kwolekr) [inline src/script/lua_api/l_mainmenu.cpp]: If you return 0 instead, you can remove this lua_pushnil(L).

**Comment by kwolekr** [comment]
I totally agree with VanessaE, it would be great if there was some kind of error message translation step.  As it is right now this just makes the same unfriendly error display in the mmdb window.  Non-technical users wouldn't understand this and may report it as a bug.

File: src/convert_json.cpp [hunk 2]
```diff
@@ -43,17 +43,21 @@ Json::Value fetchJsonValue(const std::string &url,
 	httpfetch_sync(fetch_request, fetch_result);
 
 	if (!fetch_result.succeeded) {
-		return Json::Value();
+		std::string err_msg = "HTTP request failed with code ";
+		err_msg += itos(fetch_result.response_code);
+		throw JsonFetchException(err_msg.c_str());
```
Comment (kwolekr) [inline src/convert_json.cpp]: Maybe JsonFetchException could have a http_response_code member added to it.  Perhaps there could even be a JsonFetchException::getFriendlyErrorString().

File: src/convert_json.h [hunk 3]
```diff
@@ -21,10 +21,19 @@ with this program; if not, write to the Free Software Foundation, Inc.,
...
+#include "exceptions.h"
 
 struct ModStoreMod;
 struct ModStoreModDetails;
 
+class JsonFetchException : public BaseException {
+public:
+	JsonFetchException(const char *s):
+		BaseException(s)
+	{
```
Comment (kwolekr) [inline src/convert_json.h]: Maybe JsonFetchException could have a http_response_code member added to it. Perhaps there could even be a JsonFetchException::getFriendlyErrorString().

File: src/script/lua_api/l_mainmenu.cpp [hunk 4]
```diff
@@ -409,15 +414,20 @@ int ModApiMainMenu::l_get_modstore_list(lua_State *L)
 {
 	Json::Value mods;
 	std::string url = "";
-	try{
+	try {
 		url = g_settings->get("modstore_listmods_url");
-	}
-	catch(SettingNotFoundException &e) {
+	} catch(SettingNotFoundException &e) {
```
Comment (kwolekr) [inline src/script/lua_api/l_mainmenu.cpp]: This can use Settings::getNoEx() instead of catching the exception manually.

**Comment by VanessaE** [comment]
I wasn't suggesting a "translation" step really, merely that the numeric error code be surrounded by a more user-friendly message.

File: src/script/lua_api/l_mainmenu.cpp (hunk 4 above)
Comment (est31) [inline src/script/lua_api/l_mainmenu.cpp]: Or just not catching it at all, I think that setting has a default.

**Comment by est31** [comment]
Didn't merge, as mmdb got completely disabled in d565ce95395ec514d48b86910ae333bd5bfaa951

**Pull request closed without merging**